core/drivers/tpm2/tpm2_ptp_fifo.c

```diff
@@ -0,0 +1,463 @@
+	uint32_t burst = 0;
+	uint64_t timeout_ref = 0;
+
+	if (chip->locality < 0)
+		return TPM2_ERR_INVALID_ARG;
+
+	timeout_ref = timeout_init_us(chip->timeout_a * 1000);
+	/* wait for burstcount */
+	do {
+		ops->rx32(chip, TPM2_STS(chip->locality), &burst);
+		*burstcount = (burst >> 8) & 0xFFFF;
+		if (*burstcount)
+			return TPM2_OK;
+
+		mdelay(TPM2_TIMEOUT_RETRY_MS);
+	} while (!timeout_elapsed(timeout_ref));
+
+	return TPM2_ERR_TIMEOUT;
+}
+
+enum tpm2_result tpm2_fifo_init(struct tpm2_chip *chip)
+{
+	enum tpm2_result ret = TPM2_OK;
+	const struct tpm2_ptp_phy_ops *ops = chip->phy_ops;
+	uint32_t flags = 0;
+
+	assert(tpm2_fifo_check_ops(ops));
+
+	chip->timeout_a = TPM2_TIMEOUT_A;
+	chip->timeout_b = TPM2_TIMEOUT_B;
+	chip->timeout_c = TPM2_TIMEOUT_C;
+	chip->timeout_d = TPM2_TIMEOUT_D;
+	chip->locality = -1;
+
+	/* Wait for VALID bit to be set in TPM_ACCESS */
+	ret = tpm2_fifo_wait_valid(chip, 0);
+	if (ret)
+		return TPM2_ERR_NODEV;
+
+	ret = tpm2_fifo_request_locality(chip, 0);
+	if (ret)
+		return ret;
+
+	/* disable interrupts */
+	chip->phy_ops->rx32(chip, TPM2_INT_ENABLE(chip->locality), &flags);
+	flags |= TPM2_INT_CMD_READY_INT | TPM2_INT_LOCALITY_CHANGE_INT |
+		 TPM2_INT_DATA_AVAIL_INT | TPM2_INT_STS_VALID_INT;
+	flags &= ~TPM2_GLOBAL_INT_ENABLE;
+	chip->phy_ops->tx32(chip, TPM2_INT_ENABLE(chip->locality), flags);
+
+	return tpm2_fifo_relinquish_locality(chip);
+}
+
+enum tpm2_result tpm2_fifo_end(struct tpm2_chip *chip)
+{
+	/* Cancel any command that may have been sent to TPM */
+	tpm2_fifo_set_status(chip, TPM2_STS_COMMAND_READY);
+	/* Relinquish locality if it waas requested earlier */
+	tpm2_fifo_relinquish_locality(chip);
+
+	return TPM2_OK;
+}
+
+/* Based on Figure 3 - Send Command using FIFO protocol in DD Specification */
+enum tpm2_result tpm2_fifo_send(struct tpm2_chip *chip, uint8_t *buf,
+				uint32_t len)
+{
+	enum tpm2_result ret = TPM2_OK;
+	const struct tpm2_ptp_phy_ops *ops = NULL;
+	uint32_t burstcnt = 0;
+	uint32_t sent = 0;
+	uint32_t wr_size = 0;
+	uint32_t buf_len = 0;
+	uint8_t retry_cnt = 0;
+	uint8_t status = 0;
+
+	if (!chip)
+		return TPM2_ERR_GENERIC;
+
+	ops = chip->phy_ops;
+
+	/* locality will be relinquishd in tpm2_rx() */
+	ret = tpm2_fifo_request_locality(chip, 0);
+	if (ret)
+		return ret;
+
+	while (retry_cnt < TPM2_DD_RETRY_CNT) {
+		buf_len = len;
+		sent = 0;
+		wr_size = 0;
+		burstcnt = 0;
+		status = 0;
+
+		ret = tpm2_fifo_get_status(chip, &status);
+		if (ret)
+			break;
+
+		if (!(status & TPM2_STS_COMMAND_READY)) {
+			ret = tpm2_fifo_set_status(chip,
+						   TPM2_STS_COMMAND_READY);
+			if (ret) {
+				EMSG("Previous cmd cancel failed");
+				break;
+			}
+			ret = tpm2_fifo_wait_for_status(chip,
+							TPM2_STS_COMMAND_READY,
+							chip->timeout_b,
+							&status);
+			if (ret) {
+				retry_cnt++;
+				continue;
+			}
+		}
+
+		while (buf_len > 0) {
+			ret = tpm2_fifo_get_burstcount(chip, &burstcnt);
+			if (ret) {
+				retry_cnt++;
+				continue;
```

Comment:
this while continue in the current `while()` loop, not the one above, if i'm right.

Reply:
Good catch. Thanks. Will fix this